lib/ddtrace/sampler.rb

```diff
@@ -42,8 +44,66 @@ def sample_rate=(sample_rate)
     end
 
     def sample(span)
-      span.sampled = ((span.trace_id * KNUTH_FACTOR) % Datadog::Span::MAX_ID) <= @sampling_id_threshold
       span.set_metric(SAMPLE_RATE_METRIC_KEY, @sample_rate)
+      span.sampled = ((span.trace_id * KNUTH_FACTOR) % Datadog::Span::MAX_ID) <= @sampling_id_threshold
+    end
+  end
+
+  # \RateByServiceSampler samples different services at different rates
+  class RateByServiceSampler < Sampler
+    DEFAULT_KEY = 'service:,env:'.freeze
+
+    def initialize(rate = 1.0, opts = {})
+      @env = opts.fetch(:env, Datadog.tracer.tags[:env])
+      @mutex = Mutex.new
+      @fallback = RateSampler.new(rate)
+      @sampler = { DEFAULT_KEY => @fallback }
+    end
+
+    def sample(span)
+      key = key_for(span)
+
+      @mutex.synchronize do
+        @sampler.fetch(key, @fallback).sample(span)
+      end
+    end
+
+    def update(rate_by_service)
+      @mutex.synchronize do
+        @sampler.delete_if { |key, _| key != DEFAULT_KEY && !rate_by_service.key?(key) }
+
+        rate_by_service.each do |key, rate|
+          @sampler[key] ||= RateSampler.new(rate)
+          @sampler[key].sample_rate = rate
+        end
+      end
+    end
+
+    private
+
+    def key_for(span)
+      "service:#{span.service},env:#{@env}"
+    end
+  end
+
+  # \PrioritySampler
+  class PrioritySampler
+    extend Forwardable
+
+    def initialize(opts = {})
+      @base_sampler = opts[:base_sampler] || RateSampler.new
+      @post_sampler = opts[:post_sampler] || RateByServiceSampler.new
     end
+
+    def sample(span)
+      span.sampling_priority = 0
```

Comment:
the `sampling_priority` attribute must be in the `Context` rather than `Span`. In that way you can set it with something similar to:
```ruby
span.context.sampling_priority = 2
```
And at the end when the `Context`  is finished, you just attach that metric to the root span (first element of the trace because it's a list). In Python we do: https://github.com/DataDog/dd-trace-py/blob/master/ddtrace/context.py#L148-L167